File: libs/power-log-parser/src/lib/parsers/battlegrounds-minions-sold-parser.ts
import { BlockType, CardIds, GameTag, Zone } from '@firestone-hs/reference-data';
import { ActionParser } from '../action-parser';
import { GameEventHelper, GameEventProvider } from '../game-event';
import { Action, Node, TagChange } from '../models';
import { GameState } from '../state/game-state';
import { ParserState, StateType } from '../state/parser-state';
import type { StateFacade } from '../state/state-facade';

export class BattlegroundsMinionsSoldParser implements ActionParser {
	readonly ParserName = 'BattlegroundsMinionsSoldParser';

	private GameState: GameState;
	private ParserState: ParserState;
	private StateFacade: StateFacade;

	constructor(parserState: ParserState, facade: StateFacade) {
		this.ParserState = parserState;
		this.GameState = parserState.GameState;
		this.StateFacade = facade;
	}

	AppliesOnNewNode(_node: Node, _stateType: StateType): boolean {
		return false;
	}

	AppliesOnCloseNode(node: Node, stateType: StateType): boolean {
		return (
			stateType === StateType.PowerTaskList &&
			this.StateFacade.IsBattlegrounds() &&
			node.Type === Action &&
			(node.Object as Action).Type === (BlockType.POWER as number) &&
			this.GameState.CurrentEntities.has((node.Object as Action).Entity) &&
			this.GameState.CurrentEntities.get((node.Object as Action).Entity)!.CardId ===
				CardIds.DragToSell
		);
	}

	CreateGameEventProviderFromNew(_node: Node): GameEventProvider[] | null {
		return null;
	}

	CreateGameEventProviderFromClose(node: Node): GameEventProvider[] | null {
		const action = node.Object as Action;
		const zoneTagChange = action
			.GetDataRecursive()
			.filter((data) => data instanceof TagChange)
			.map((data) => data as unknown as TagChange)
			.find(
				(tag) =>
					tag.Name === (GameTag.ZONE as number) && tag.Value === (Zone.SETASIDE as number),
			);
		if (zoneTagChange == null) {
			return null;
		}

		const soldMinion = this.GameState.CurrentEntities.get(zoneTagChange.Entity)!;
		const cardId = soldMinion.CardId;
		const controller = soldMinion.GetEffectiveController();
		return [
			GameEventProvider.Create(
				(node.Object as Action).TimeStamp,
				'BATTLEGROUNDS_MINION_SOLD',
				GameEventHelper.CreateProvider(
					'BATTLEGROUNDS_MINION_SOLD',
					cardId,
					controller,
					soldMinion.Entity,
					this.StateFacade,
					null,
				),
				true,
				node,
			),
		];
	}
}

File: tools/linux-helpers/hs-converter/HearthstoneReplays/Events/Parsers/BattlegroundsMinionsSoldParser.cs
using HearthstoneReplays.Parser;
using HearthstoneReplays.Parser.ReplayData;
using HearthstoneReplays.Parser.ReplayData.GameActions;
using System.Linq;
using HearthstoneReplays.Enums;
using HearthstoneReplays.Parser.ReplayData.Entities;
using System.Collections.Generic;
using static HearthstoneReplays.Events.CardIds;
using System;
using Action = HearthstoneReplays.Parser.ReplayData.GameActions.Action;

namespace HearthstoneReplays.Events.Parsers
{
    public class BattlegroundsMinionsSoldParser : ActionParser
    {
        private GameState GameState { get; set; }
        private ParserState ParserState { get; set; }
        private StateFacade StateFacade { get; set; }

        public BattlegroundsMinionsSoldParser(ParserState ParserState, StateFacade facade)
        {
            this.ParserState = ParserState;
            this.GameState = ParserState.GameState;
            this.StateFacade = facade;
        }

        public bool AppliesOnNewNode(Node node, StateType stateType)
        {
            return false;
        }

        public bool AppliesOnCloseNode(Node node, StateType stateType)
        {
            return stateType == StateType.PowerTaskList
                && StateFacade.IsBattlegrounds()
                && node.Type == typeof(Action)
                && (node.Object as Action).Type == (int)BlockType.POWER
                && GameState.CurrentEntities.ContainsKey((node.Object as Action).Entity)
                && GameState.CurrentEntities[((node.Object as Action).Entity)].CardId == CardIds.DragToSell;
        }

        public List<GameEventProvider> CreateGameEventProviderFromNew(Node node)
        {
            return null;
        }

        public List<GameEventProvider> CreateGameEventProviderFromClose(Node node)
        {
            var action = node.Object as Action;
            // Because Sellemental triggers another action, and the zone change is in that new action
            var zoneTagChange = action.GetDataRecursive()
                .Where(data => data is TagChange)
                .Select(data => data as TagChange)
                .Where(tag => tag.Name == (int)GameTag.ZONE && tag.Value == (int)Zone.SETASIDE)
                .FirstOrDefault();
            if (zoneTagChange == null)
            {
                return null;
            }

            var soldMinion = GameState.CurrentEntities[zoneTagChange.Entity];
            var cardId = soldMinion.CardId;
            var controller = soldMinion.GetEffectiveController();
            return new List<GameEventProvider> { GameEventProvider.Create(
                    (node.Object as Action).TimeStamp,
                     "BATTLEGROUNDS_MINION_SOLD",
                    GameEvent.CreateProvider(
                        "BATTLEGROUNDS_MINION_SOLD",
                        cardId,
                        controller,
                        soldMinion.Entity,
                        StateFacade,
                        null),
                    true,
                    node)
                };
        }
    }
}
